internal/subscriber/core.go

```diff
@@ -39,36 +40,28 @@ func (c *Core) NewSubscriber(ctx context.Context,
 	ackCh chan *AckMessage,
 	modAckCh chan *ModAckMessage) (ISubscriber, error) {
 
-	topic := messagebroker.NormalizeTopicName(subscription.Topic)
-	retryTopic := messagebroker.NormalizeTopicName(subscription.GetRetryTopic())
-	dlqTopic := messagebroker.NormalizeTopicName(subscription.GetDeadLetterTopic())
-
 	groupID := subscription.Name
 
-	consumer, err := c.bs.GetConsumer(ctx, messagebroker.ConsumerClientOptions{Topics: []string{topic, retryTopic}, GroupID: groupID, GroupInstanceID: subscriberID})
+	consumer, err := c.bs.GetConsumer(ctx, messagebroker.ConsumerClientOptions{Topics: []string{subscription.Topic, subscription.GetRetryTopic()}, GroupID: groupID, GroupInstanceID: subscriberID})
 	if err != nil {
 		logger.Ctx(ctx).Errorw("subscriber: failed to create consumer", "error", err.Error())
 		return nil, err
 	}
 
-	retryProducer, err := c.bs.GetProducer(ctx, messagebroker.ProducerClientOptions{Topic: retryTopic, TimeoutMs: 500})
-	if err != nil {
-		logger.Ctx(ctx).Errorw("subscriber: failed to create retry producer", "error", err.Error())
-		return nil, err
-	}
-
-	dlqProducer, err := c.bs.GetProducer(ctx, messagebroker.ProducerClientOptions{Topic: dlqTopic, TimeoutMs: 500})
-	if err != nil {
-		logger.Ctx(ctx).Errorw("subscriber: failed to create dlq producer", "error", err.Error())
-		return nil, err
+	subsCtx, cancelFunc := context.WithCancel(ctx)
+	// using the subscriber ctx for retrier as well. This way when the ctx() for subscribers is done,
+	// all the delay-consumers spawned within retrier would also get marked as done.
+	var retrier retry.IRetrier
+	if subscription.RetryPolicy != nil && subscription.DeadLetterPolicy != nil {
```

Comment:
these should never be null, can we handle it gracefully for older subscriptions.
i.e  subscription.GetRetryPolicy() returns default policy if its not set and same for deadletterpolicy

once migrated we can remove the nulll checks or let them be there for safety

Reply:
Doing a fallback only on default Retry and PL policies will not suffice as we need to have the delay topics created as well for retries to work. Ideally, the migration step should take care of old subscriptions and all new subscriptions after this PR should have non-nil values.